### PR TITLE
text sequence classification using Glove and RNN/LSTMs

### PR DESCRIPTION
**Pull request by thammegowda**
Summary: 
+ Added a dataset reader for feeding mini batches to DL4J's network
+  Golve embeddings to vectorize text using Stanford NLP group's pre trained Glove vectors
+ A tiny (2 layer) classifier based on LSTMs
+ All of the above are written hoping to reuse them for other multi class text classifiers. We can customize these easily:
  + number of classes 
  + vector embeddings
  + vocabulary size
  + number of LSTM cells, batch size, etc..

### Known issues:
 When learning rate is too low or too high, the gradients quickly bounce to Infinity or NaN.
Based on the dataset and vectors,  the `-lr` parameter should be tuned.

### Datasets 
```
# Download pre trained Glo-ves (this is a large file)
 wget http://nlp.stanford.edu/data/glove.6B.zip
unzip glove.6B.zip -d glove.6B

# Download dataset
 wget http://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz
tar xzf aclImdb_v1.tar.gz
```

Note: try it out on smaller datasets first.  Suggestion: create `aclImdb/train-lite` and `aclImdb/test-lite` with 1500 positive and 1500 negative examples each from IMDB dataset. 

### Data Set Organization:
In general, organize the directory as follows:
```
data-dir/
    + train/
    |  +- label1 /
    |  |    +- example11.txt
    |  |    +- example12.txt
    |  |    +- example13.txt
    |  |    +- .....
    |  +- label2 /
    |  |    +- example21.txt
    |  |    +- .....
    |  +- labelN /
    |       +- exampleN1.txt
    |       +- .....
    + test/
         + label1/
              +- ........
```
Note:  IMDB large dataset is already shipped in this format, just reduce file count for quicker testing

```
alias RUN="mvn compile exec:java -Dexec.mainClass=opennlp.tools.dl.GloveRNNTextClassifier"
```

## Train 
```bash
RUN -Dexec.args="-glovesPath glove.6B/glove.6B.100d.txt  \
    -labels pos neg -modelPath imdb-sentimodel.dat \
    -trainDir=aclImdb/train-lite -lr 0.001"
```

## Predict 
```bash
RUN  -Dexec.args="-glovesPath glove.6B/glove.6B.100d.txt \
      -labels pos neg -modelPath imdb-sentimodel.dat \
       -files /aclImdb/test/pos/1_10.txt datasets/aclImdb/test/neg/1_3.txt"
```

---
## CLI Arguments and Default values to GloveRNNTextClassifier:
```
 -batchSize N       : Number of examples in minibatch. Applicable for training
                      only. (default: 128)
 -files STRING[]    : File paths (separated by space) to predict using the
                      model.
 -glovesPath VAL    : Path to GloVe vectors file. Download and unzip from
                      https://nlp.stanford.edu/projects/glove/
 -labels STRING[]   : Names of targets or labels separated by spaces. The order
                      of labels matters. Make sure to use the same sequence for
                      training and predicting. Also, these names should match
                      subdirectory names of -trainDir and -validDir when those
                      are applicable.
                       Example -labels pos neg
 -lr (-learnRate) N : Learning Rate. Adjust it when the scores bounce to NaN or
                      Infinity. (default: 0.002)
 -maxSeqLen N       : Max Sequence Length. Sequences longer than this will be
                      truncated (default: 256)
 -modelPath VAL     : Path to model file. This will be used for serializing the
                      model after the training phase.and also the model will be
                      restored from here for prediction
 -nEpochs N         : Number of epochs (i.e. full passes over the training
                      data) to train on. Applicable for training only.
                      (default: 2)
 -nRNNUnits N       : Number of RNN cells to use. Applicable for training only.
                      (default: 128)
 -trainDir VAL      : Path to train data directory. Optional. Setting this
                      value will take the system to training mode.
 -validDir VAL      : Path to validation data directory. Optional. Applicable
                      only when -trainDir is set.
 -vocabSize N       : Vocabulary Size. (default: 20000)

```

## References:
+ Glove - https://nlp.stanford.edu/projects/glove/
+ RNN and LSTM support in DL4j - https://deeplearning4j.org/usingrnns

### CODE REVIEW

**Comment by tteofili** [comment]
thanks a lot @thammegowda for this PR, it looks very interesting, I think it makes sense to check whether we can make the text classifier implement name finder or doccat openNLP APIs.

**Comment by kottmann** [comment]
We have some similar work already for the name finder, like @tteofili says it would be great if we could implement doccat for the text classification part. This will allow us to run all the existing tooling, e.g. for training or evaluation.

Also we had some issues with memory management and dl4j, when this runs on GPU it seems to me the wrong approach to load the entire GLVOE vectors into the GPU memory.

I suggest you come over to the dev list and we discuss this a bit, there are also other people who are very interested in getting this moving.

**Comment by thammegowda** [comment]
Thanks @tteofili and @kottmann for the comments.

> it would be great if we could implement doccat for the text classification part.

Got it. I will plan to make it compatible with Document Categorizer interface/baseclass/contract. Then we can re use tools for evaluation

> Also we had some issues with memory management and dl4j, when this runs on GPU it seems to me the wrong approach to load the entire GLVOE vectors into the GPU memory. Also dl4j ships with support to load the GLOVE vectors similar to how you do it.

+1. I tried the [Word2VecSentimentRNN example](
https://github.com/deeplearning4j/dl4j-examples/blob/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/recurrent/word2vecsentiment/Word2VecSentimentRNN.java) which uses word vectors from https://code.google.com/archive/p/word2vec/ 
Even I also hit the Out of Memory issue. I couldn't find a way to select only topN frequent words without messing with its code.
Also I have been hearing good words about Glove from Stanford NLP group, so I thought of writing importer to use prebuilt Glove vectors. (Is there a way to do it in DL4J already? I was not knowing it!)
The `-vocabSize` CLI argument can be used to control the number of top words to use. (rest all words will be skipped for now, but may be treated as UNKNOWN by extending this.)

> I suggest you come over to the dev list and we discuss this a bit, there are also other people who are very interested in getting this moving.

:+1: :100:  Coming now

**Comment by agibsonccc** [comment]
Hey folks - I believe @kottmann was messing with this before. Pretrained/broad word vectors are a bit of a problem due to sheer vocab size For nd4j, we are thinking about some sort of in memory compression for sparse access patterns like word embeddings. If you have any restrictions, I would invite you to file issues.

https://github.com/deeplearning4j/deeplearning4j/issues


We have helped other apache projects in a similar manner. We would like to make sure you have a good experience. Please let us know what we can do to help or improve.

**Comment by kottmann** [comment]
@agibsonccc let me spend some more time with the Named Entity Recognition stuff I did and change it so it doesn't use nd4j to store the word vectors. Let me put that in a PR and maybe have you review it.

**Comment by thammegowda** [comment]
Work in progress:
1. Implemented DocumentCategorizer interface
2. Separated the code into Trainer, Model, Predictor 
3. Model can be serialized (with network topology, weights, glove embedding vectors, few other params used for training) into a single zip file and then it can be restored by predictor tool

How to test at this state:
```
alias TRAIN="mvn compile exec:java -Dexec.mainClass=opennlp.tools.dl.NeuralDocCatTrainer"
alias PREDICT="mvn compile exec:java -Dexec.mainClass=opennlp.tools.dl.NeuralDocCat"
```

```
$ TRAIN

 -batchSize N       : Number of examples in minibatch (default: 128)
 -glovesPath VAL    : Path to GloVe vectors file. Download and unzip from
                      https://nlp.stanford.edu/projects/glove/
 -labels STRING[]   : Names of targets or labels separated by spaces. The order
                      of labels matters. Make sure to use the same sequence for
                      training and predicting. Also, these names should match
                      subdirectory names of -trainDir and -validDir when those
                      are applicable.
                       Example -labels pos neg
 -lr (-learnRate) N : Learning Rate. Adjust it when the scores bounce to NaN or
                      Infinity. (default: 0.002)
 -maxSeqLen N       : Max Sequence Length. Sequences longer than this will be
                      truncated (default: 256)
 -modelPath VAL     : Path to model file. This will be used for serializing the
                      model after the training phase.
 -nEpochs N         : Number of epochs (i.e. full passes over the training
                      data) to train on. Applicable for training only.
                      (default: 2)
 -nRNNUnits N       : Number of RNN cells to use. (default: 128)
 -trainDir VAL      : Path to train data directory. Setting this value will
                      take the system to training mode.
 -validDir VAL      : Path to validation data directory. Optional.
 -vocabSize N       : Vocabulary Size. (default: 20000)
```

```
$ PREDICT 
 -files FILE : One or more document paths whose category is to be predicted by
               the model
 -model VAL  : Path to NeuralDocCatModel stored file

```

**Comment by chrismattmann** [comment]
looking good @thammegowda

**Comment by tteofili** [comment]
+1 thanks @thammegowda for the improbements

**Comment by kottmann** [comment]
+1

**Comment by chrismattmann** [comment]
+1 to merge

**Comment by tteofili** [comment]
We'll need to clarify if:
* nd4j loading GloVe embeddings is problematic memory wise
* GPU support can actually improve performance
All the above maybe with the help of @agibsonccc :)
Probably it'd be useful to also add CLI support (as noted in a TODO).
That being said, I think we can merge now as @thammegowda has provided the Doccat API implementation.
Above issues and additional improvements can be addressed in separate PRs.

**Comment by chrismattmann** [comment]
yay, thanks @thammegowda and @tteofili

**Comment by thammegowda** [comment]
Thanks for merging :-) 

PR Merged

**Comment by chrismattmann** [comment]
nice @thammegowda

**Comment by agibsonccc** [comment]
Great job all! 

@tteofili yes it might be depending on the vocab size. The default data type is float. My best suggestion to you would be to  engage @raver119 on some ideas here.

GPU support is another thing that *should* help. We have found it helps with LSTMs at least. Would be worth looking in to.

We will be doing a release in the next week or so with workspaces improvements: http://deeplearnin4j.org/workspaces

You can try it out on oss-sonatype snapshots  to see what it does as well.
